mne/inverse_sparse/mxne_optim.py

```diff
@@ -768,6 +768,115 @@ def norm_l1_tf(Z, shape, n_orient):
+
+    Parameters
+    ----------
+    Y : array, shape (n_atoms,)
+        The input data.
+    l2_ratio : float between 0 and 1
+        Tradeoff between L2 and L1 regularization.
+
+    Returns
+    -------
+    nu : float
+        The value of the dual norm evaluated at Y.
+
+    References
+    ----------
+    .. [1] E. Ndiaye, O. Fercoq, A. Gramfort, J. Salmon,
+       "GAP Safe Screening Rules for Sparse-Group Lasso", Advances in Neural
+       Information Processing Systems (NIPS), 2016.
+    """
+    # since the solution is invariant to flipped signs in Y, all entries
+    # of Y are assumed positive
+    norm_inf_Y = np.max(Y)
+    if l2_ratio == 0:
+        # dual norm of L1 is Linf
+        return norm_inf_Y
+    elif l2_ratio == 1.:
+        # dual norm of L2 is L2
+        return np.sqrt((Y ** 2).sum())  # TODO Y is always real
+
+    if norm_inf_Y == 0.:
+        return 0.
+
+    K = (Y > (1. - l2_ratio) * norm_inf_Y).sum()
+    if K == 1:
+        return norm_inf_Y
+
+    # K largest values of Y, in decreasing order, without sorting Y:
+    Y = np.sort(Y[np.argpartition(Y, len(Y) - K)[len(Y) - K:]])[::-1]
+
+    p_sum = 0.  # partial sum
+    p_sum_2 = 0.  # partial sum of squares
+    lower = 0.
+
+    for k in range(K - 1):
+        p_sum += Y[k]
+        p_sum_2 += Y[k] ** 2
+        upper = p_sum_2 / Y[k + 1] ** 2 - 2. * p_sum / Y[k + 1] + k + 1
+        if lower <= l2_ratio ** 2 / (1. - l2_ratio) ** 2 < upper:
+            j = k + 1
+            break
+        lower = upper
+    else:
+        j = K
+        p_sum += Y[K - 1]
+        p_sum_2 += Y[K - 1] ** 2
+
+    denom = (1. - l2_ratio) ** 2 * j - l2_ratio ** 2
+    if np.abs(denom) < 1e-10:
+        return p_sum_2 / 2. * (1. - l2_ratio) * p_sum
+    else:
+        # TODO prove that we cannot get a negative value in the sqrt
+        delta = ((1. - l2_ratio) * p_sum) ** 2 - p_sum_2 * denom
+        return ((1. - l2_ratio) * p_sum - np.sqrt(delta)) / denom
+
+
+def norm_epsilon_inf(G, R, phi, l2_ratio, n_orient):
+    """epsilon-inf norm of phi(np.dot(G.T, R)).
+
+    Parameters
+    ----------
+    G : array, shape (n_sensors, n_sources)
+        Gain matrix a.k.a. lead field.
+    R : array, shape (n_sensors, n_times)
+        Residuals.
+    phi : instance of _Phi
+        The TF operator.
+    l2_ratio : float between 0 and 1
+        Parameter controlling the tradeoff between L21 and L1 regularization.
+    n_orient : int
+        Number of dipoles per location (typically 1 or 3).
+
+    Returns
+    -------
+    nu : float
+        The maximum value of the epsilon norms over groups of n_orient dipoles.
+    """
+    n_positions = G.shape[1] // n_orient
+    GTRPhi = np.abs(phi(np.dot(G.T, R))) ** 2
+    # norm over orientations:
+    GTRPhi = np.sqrt(np.sum(GTRPhi.reshape((n_orient, -1), order='F'),
+                     axis=0)).reshape((n_positions, -1), order='F')
+
+    nu = 0.
+    for idx in range(n_positions):
```

Comment:
no way to vertorize this? a loop over 8000 sources for each dual gap check will be costly :(